doc/threshold/SONiC Threshold feature spec.md

```diff
@@ -0,0 +1,589 @@
+| ThresholdMgr             | Threshold manager                   |
+
+# 1 Feature Overview
+The threshold feature allows configuration of a threshold on supported buffers in ingress and egress. A threshold breach notification (entry update in COUNTERS\_DB) is generated on breach of the configured buffer threshold in ASIC.   
+
+## 1.1 Requirements
+### 1.1.1 Functional Requirements
+
+1.0 Threshold feature allows the user to configure a threshold on a buffer (i.e buffer usage) and record all threshold breach events for that buffer.
+
+1.1 On breach of the configured threshold in ASIC, a breach event entry is added to COUNTERS\_DB. User can list out the breach event data through a CLI command. In addition, the breach entries in COUNTERS\_DB are also available to SONiC telemetry feature as well.
+
+1.2 The threshold feature can be used to track and be notified of congestion events in the network real-time.
+
+1.3 Monitoring of buffer usage can be achieved via the watermark feature available in SONiC.
+
+1.4 Along with the watermark feature, the threshold feature provides data that can be used to design applications to monitor MMU buffer usage.  
+
+2.0.0 The threshold for a buffer must be configured in terms of percentage of buffer allocation.
+ 
+2.0.1 The configured threshold can be cleared at any point by the user.
+
+2.0.2 The supported counters for threshold feature are listed below.
+
+2.0.2.1 SAI\_INGRESS\_PRIORITY\_GROUP\_STAT\_SHARED\_WATERMARK\_BYTES -- Current ingress priority-group shared buffer usage in bytes.
+
+2.0.2.2 SAI\_INGRESS\_PRIORITY\_GROUP\_STAT\_XOFF\_ROOM\_WATERMARK\_BYTES -- Current ingress priority-group headroom buffer usage in bytes.
+
+2.0.2.3 SAI\_QUEUE\_STAT\_SHARED\_WATERMARK\_BYTES -- Current egress queue buffer usage in bytes.
+
+2.0.3 The counters listed in 2.0.2.1, 2.0.2.2 and 2.0.2.3 are to be supported for all ports on SONiC.
+
+2.0.4 The threshold can be configured on one or more of the 8 priority-groups of a port.
+
+2.0.5 The threshold can be configured on one or more of the unicast/multicast queues of a port.
+
+2.0.6 On breach of the configured threshold, the breach information is written into  THRESHOLD\_BREACH\_TABLE of COUNTERS\_DB. 
+
+2.0.6.1 The breach entry must indicate the counter on which the threshold breached.
+
+2.0.6.2 The breach entry must indicate the port associated with the counter on which the breach occurred.
+
+2.0.6.3 The breach entry must indicate the counter index of the port (ref 2.0.4, 2.0.5) on which the breach occurred i.e the priority-group index for 2.0.2.1, 2.0.2.2 and queue index and type for 2.0.2.3. 
+
+2.0.6.5 The breach entry must contain the buffer usage in terms of percentage at the time of breach.
+
+2.0.6. The breach entry must contain the absolute buffer usage in bytes at the time of breach.
+
+2.0.6.4 The breach entry must contain the time stamp of the breach event. The ASIC provides the time stamp of breach event via the breach event protobuf.
```

Comment:
Please fix the item numberings in this section.

Reply:
Will do ASAP.

Reply:
Comment addressed and change uploaded.